### ISSUE TITLE
Add possibility to consistently get object property name

### ISSUE THREAD

**Issue by DisDis** [dart-lang/sdk#31255]
We need to get object property name in cases like Angular 2 ngOnChanges (when you need to compare key from changes map with object property name):

ngOnChanges(Map<String, SimpleChange> changes);
It would be great to have something like nameof in C#, so that we could get property name and don't be afraid that someone will rename property.

`nameof(obj.propertyName); // 'propertyName'`

**Comment by matanlurey** [comment]
Considering that you can't _use_ this information (i.e. there is no reflective support), what would this accomplish?

**Comment by lrhn** [comment]
You can use it in `noSuchMethod`.
It might even help you with the named arguments of `Function.apply`, but only if you are somehow allowed to denote a named parameter of a function signature.

It's a good idea. I wouldn't use a macro like `nameof` (Dart doesn't have precompiler macros otherwise), but I would definitely like something that takes qualified name in scope and provides the symbol for that name. It should be recognizable in renaming and other refactorings.

We can't use `#foo` since that already works for things that are not in scope.
So, for the giggles, let's try `##foo`. It would have a grammar like a symbol, but be interpreted differently:
```dart
import "self.dart" as self;
abstract class C {
  static foo({bar});
  baz({qux});
  get toto;
}
main() {
  var c = ##C;
  var cFoo = ##C.foo;
  var cFooBar = ##C.foo.bar;
  var cFooBar2 = ##self.C.foo.bar;
  print(identical(cFooBar, cFooBar2));  // true, yey!
  // var wrong = ##C.biff;  // Invalid, no biff declared in C.
}
```

Other possible, but probably too odd, syntaxes: `#foo.bar#`, `#:foo.bar`, ...
No great syntax so far :(

See also #30518,

**Comment by ranquild** [comment]
@matanlurey `if (changes.containsKey(nameof(field)) ...`.

**Comment by matanlurey** [comment]
At least for that particular API, we decided it was probably wrong.

I don't know if I'd push for a language feature to fix the API we created (personally).

**Comment by DisDis** [comment]
In our big project. Main priority is code support. We have 'toName', 'toSymbol' an analog of nameof. It is made using a transformer and the special code for run vm test.
We very often use this for:
Deserialization/Serialization, Angular.Dart, noSuchMethod

```
// Deserialization
class AccountDeserializer extends EntityDeserializer {
  static final _rightsSymbol = toSymbol((Account o) => o.rights);
  static final List<DeserializeField> _fields = [
    new DeserializeSingleField("isOwner", toSymbol((Account o) => o.isOwner)),
    new DeserializeSingleField("isAdmin", toSymbol((Account o) => o.isAdmin)),
....

// Angular
ngOnChanges(Map<String, SimpleChange> changes){
   if (changes.containsKey(nameof(field)) ....
}

```

Interface
```
toSymbols([o.field1, o.field2]);// => [#field1,#field2]
toSymbol(o.field1);// => #field1
toSymbols((Task o)=>[o.field1, o.field2]);// => [#field1,#field2]
toSymbol((Task o)=>o.field1);// => #field1

toNames([o.field1, o.field2]);// => ['field1','field2']
toName(o.field1);// => 'field1'
toNames((Task o)=>[o.field1, o.field2]);// => ['field1','field2']
toName((Task o)=>o.field1);// => 'field1'
```

All this is directed to ensure that the code is connected at all parts. Developers should receive a real "Find usages". Developers should safely refactor the code. Also in our company, literals are forbidden. Since literals are potential errors.
Now the transformer has become a problem, as we move to build. 'nameof' solves the problems at the language level.

@matanlurey @lrhn

**Comment by jodinathan** [comment]
this would help me so much.

All my models are like:

```
class SomeClass extends Model {
  int _myProp;

  int get myProp => _myProp;
  set myProp(int x) {
    this.modified('myProp', _myProp, _myProp = x);
  }
}
```
would be amazing to do something like: 

```
class SomeClass extends Model {
  int _myProp;

  set myProp(int x) {
    this.modified(##SomeClass.myProp, _myProp, _myProp = x);
    // becomes
    this.modified('myProp', _myProp, _myProp = x);
  }
}
```

**Comment by kevmoo** [comment]
@lrhn @munificent @leafpetersen – is this more of a language request?

I've been hitting this a lot in the last few days – would make a lot of folks very happy!

**Comment by munificent** [comment]
Yes, this sounds like a language request to me. Want to move it over to the language repo?

**Comment by kevmoo** [comment]
> Yes, this sounds like a language request to me. Want to move it over to the language repo?

Sure!
